dynamic_rest/filters.py

```diff
@@ -347,21 +347,50 @@ def _build_prefetch_queryset(
 
     def _add_internal_prefetches(
         self,
+        model,
         prefetches,
         requirements
     ):
         """Add internal (required) prefetches to a prefetch dictionary."""
-        paths = requirements.get_paths()
-        for path in paths:
-            # Remove last segment, which indicates a field name or wildcard.
-            # For example, {model_a : {model_b : {field_c}}
-            # should be prefetched as a__b
-            prefetch_path = path[:-1]
-            key = '__'.join(prefetch_path)
-            if key:
-                prefetches[key] = key
+
+        for source, remainder in six.iteritems(requirements):
+            if not remainder or isinstance(remainder, six.string_types):
+                # no further requirements to prefetch
+                continue
+
+            related_field = get_model_field(model, source)
+            related_model = related_field.related_model
+
+            if related_model is None:
+                # GenericForeignKey has no related_model
+                # in this case, prefetch the generic relationship itself
+                prefetches[source] = source
```

Comment:
@ryochiji is this consistent with your understanding of prefetching + generic relationships? It made the query-counting tests pass and makes sense to me, but I'm not as familiar with the GFK stuff.